server/org.eclipse.emfcloud.ecore.glsp/src/main/java/org/eclipse/emfcloud/ecore/glsp/model/EcoreModelServerAccess.java

```diff
@@ -98,78 +132,344 @@ private EPackage getEPackage(EcoreModelState modelState) {
 		return facade.getEPackage();
 	}
 
-	public boolean addEClassifier(EcoreModelState modelState, EClassifier newEClassifier) {
-		return this.add(modelState, getEPackage(modelState), EcorePackage.eINSTANCE.getEPackage_EClassifiers(), newEClassifier);
+	private Diagram getDiagram(EcoreModelState modelState) {
+		EcoreFacade facade = EcoreModelState.getEcoreFacade(modelState);
+		return facade.getDiagram();
+	}
+
+	public boolean addEClassifier(EcoreModelState modelState, EClassifier newEClassifier, Shape newShape) {
+		CCompoundCommand compoundCommand = CCommandFactory.eINSTANCE.createCompoundCommand();
+		compoundCommand.setType(CommandKind.COMPOUND);
+		try {
+			Command addEClassifier = createAddCommand(modelState, getEPackage(modelState),
+					EcorePackage.Literals.EPACKAGE__ECLASSIFIERS, newEClassifier);
+			Command addShape = createAddCommand(modelState, getDiagram(modelState),
+					EnotationPackage.Literals.DIAGRAM__ELEMENTS, newShape);
+
+			compoundCommand.getCommands().add(getCommandCodec().encode(addEClassifier));
+			compoundCommand.getCommands().add(getCommandCodec().encode(addShape));
+		} catch (EncodingException e) {
+			return false;
+		}
+		return this.editCompound(compoundCommand);
+	}
+
+	public boolean addEReference(EcoreModelState modelState, EReference newEReference, EClassifier source,
+			Edge newEdge) {
+		CCompoundCommand compoundCommand = CCommandFactory.eINSTANCE.createCompoundCommand();
+		compoundCommand.setType(CommandKind.COMPOUND);
+		try {
+			Command addEReference = createAddCommand(modelState, source,
+					EcorePackage.Literals.ECLASS__ESTRUCTURAL_FEATURES, newEReference);
+			Command addEdge = createAddCommand(modelState, getDiagram(modelState),
+					EnotationPackage.Literals.DIAGRAM__ELEMENTS, newEdge);
+
+			compoundCommand.getCommands().add(getCommandCodec().encode(addEReference));
+			compoundCommand.getCommands().add(getCommandCodec().encode(addEdge));
+		} catch (EncodingException e) {
+			return false;
+		}
+		return this.editCompound(compoundCommand);
 	}
-	
-	public boolean addSuperType(EcoreModelState modelState, EClassifier superType, EClassifier parent) {
-		return this.add(modelState, parent, EcorePackage.eINSTANCE.getEClass_ESuperTypes(), superType);
+
+	public boolean addEReferenceBidirectional(EcoreModelState modelState, EReference newEReference,
```

Comment:
Bidirectional References should be represented with a single Edge (The Edge represents the 2 EReferences and their "opposite" relationship). Currently, we have 2 Overlapping Edges:

![image](https://user-images.githubusercontent.com/16717379/108481772-dfc8b400-7298-11eb-81f6-15b165259c2f.png)

Also, the directions are incorrect:

- For simple EReferences, we should have 1 Edge with an Arrow towards the Type
- For bidirectional EReferences, we should have 1 Edge without any arrows
- For Containments and bidirectional Containments, we should have 1 Edge, with a diamond on the Container side (And maybe an arrow on the opposite side to distinguish the two cases; we should check and remain consistent with Ecore Tools)

Reply:
Yes I agree, the bidirectional cases need to be adjusted. We already have open issues for that compositions and references: #14 and #15 .
I will add your comments there as well.